R/verb-joins.R

```diff
@@ -237,90 +236,216 @@ add_join <- function(x, y, type, by = NULL, sql_on = NULL, copy = FALSE,
     copy = copy,
     indexes = if (auto_index) list(by$y)
   )
-
   suffix <- suffix %||% sql_join_suffix(x$src$con, suffix)
-  vars <- join_vars(op_vars(x), op_vars(y), type = type, by = by, suffix = suffix, call = call)
-
-  inlined_select_list <- inline_select_in_join(x, y, vars, by)
-  vars <- inlined_select_list$vars
-  by <- inlined_select_list$by
-
-  # the table alias can only be determined after `select()` was inlined.
-  # This works even though `by` is used in `inline_select_in_join()` and updated
-  # because this does not touch `by$x_as` and `by$y_as`.
-  by[c("x_as", "y_as")] <- check_join_as(
-    x_as,
-    inlined_select_list$x,
-    y_as,
-    inlined_select_list$y,
-    sql_on = sql_on,
-    call = call
+  suffix <- check_suffix(suffix, call)
+  na_matches <- arg_match(na_matches, c("na", "never"), error_call = call)
+
+  join_alias <- check_join_alias(x_as, y_as, sql_on, call)
+
+
+  inline_result <- join_inline_select(x$lazy_query, by$x, by$on)
+  x_lq <- inline_result$lq
+  x_vars <- inline_result$vars
+  by_x_org <- by$x
+  by$x <- inline_result$by
+
+  new_query <- join_needs_new_query(x$lazy_query, join_alias, type)
+  if (new_query) {
+    x_join_vars <- tibble(
+      name = op_vars(x),
+      table = list(1L),
```

Comment:
At this point we don't know the names used for the tables in the multi join query. So, `table` uses an integer to create a reference to the table. It must be a list to support `full_join()`.
Also `vars` must be a list to support `full_join()` together with inlining `select()`. This is a bit ugly but the alternatives I tried caused way more trouble.